iroh/src/magicsock/transports/ip.rs

```diff
@@ -0,0 +1,157 @@
+use std::{
+    io,
+    net::SocketAddr,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use n0_watcher::{Watchable, Watcher};
+use netwatch::UdpSocket;
+use tracing::trace;
+
+use super::{Addr, Transmit};
+use crate::metrics::MagicsockMetrics;
+
+#[derive(Clone, Debug)]
+pub(crate) struct IpTransport {
+    bind_addr: SocketAddr,
+    socket: Arc<UdpSocket>,
+    local_addr: Watchable<SocketAddr>,
+    metrics: Arc<MagicsockMetrics>,
+}
+
+impl IpTransport {
+    pub(crate) fn new(
+        bind_addr: SocketAddr,
+        socket: Arc<UdpSocket>,
+        metrics: Arc<MagicsockMetrics>,
+    ) -> Self {
+        // Currently gets updated on manual rebind
+        // TODO: update when UdpSocket under the hood rebinds automatically
+        let local_addr = Watchable::new(socket.local_addr().expect("invalid socket"));
+
+        Self {
+            bind_addr,
+            socket,
+            local_addr,
+            metrics,
+        }
+    }
+
+    pub(super) fn create_io_poller(&self) -> Arc<UdpSocket> {
+        self.socket.clone()
+    }
+
+    pub(super) fn poll_send(
+        &self,
+        destination: SocketAddr,
+        transmit: &Transmit<'_>,
+    ) -> Poll<io::Result<()>> {
+        trace!("sending to {}", destination);
+        let total_bytes = transmit.contents.len() as u64;
+        let res = self.socket.try_send_quinn(&quinn_udp::Transmit {
+            destination,
+            ecn: transmit.ecn,
+            contents: transmit.contents,
+            segment_size: transmit.segment_size,
+            src_ip: transmit
+                .src_ip
+                .clone()
+                .map(|a| a.try_into().expect("invalid src_ip")),
+        });
```

Comment:
Is it possible at all to keep panic-generating code out of this bit of the hot path?  I assume this has to deal with a relay addr appearing where an IP one is needed for this socket?

I could also imagine a situation where someone binds two IPv4 addresses (or IPv6, whatever) and they're both on different network interfaces, which with their own subnet.  You wouldn't be able to send with the wrong IP address in that case.  I think you're right now just bumping this into the kernel which would probably just return an error instead of sending it.  So maybe no need to worry about IP here.

Anyway, `quinn_udp::Transmit` says `src_ip: Option<IpAddr>`.  So one option would be to log a warning and turn this into `None` if you got something crazy.

Reply:
fixed